Add `error_extensionf` in the API to ease the process of embedding errors

panglesd opened the pull request:
This is part of the #314 tasks.

The examples are updated to embed errors using the new function, instead of raising located errors.

pitag-ha reviewed (commented):
Perfect, LGTM. Thanks, @panglesd! Could you add a changelog entry?

pitag-ha commented:
Oh, and one more thing: we should also edit the documentation of [`Location.raise_errorf`](https://github.com/ocaml-ppx/ppxlib/blob/main/src/location.mli#L27-L290) to point out that that function should be avoided and `error_extensionf` should be used instead. Also, the sentence "The exception is caught by the driver and handled appropriately" is very confusing, if not wrong. What do you think? Do you want to update that here or rather in a separate PR?

panglesd commented:
Sure! I added a changelog entry and updated the doc.

When this PR and the #318 PR are merged, there should be a link from the API of those functions to the relevant part of the manual. The maintenance of this link would be eased if the manual was written using `odoc` :wink:

pitag-ha commented on CHANGES.md:
```diff
@@ -19,6 +19,8 @@
   consider a type declaration recursive if the type appeared inside an attribute
   payload (#299, @NathanReb)
 
+- Added `error_extensionf` function to the `Locate` module (#316, @panglesd)
```
```suggestion
- Added `error_extensionf` function to the `Location` module (#316, @panglesd)
```

Also, similarly to [here](https://github.com/ocaml-ppx/ppxlib/pull/315#discussion_r797739592): would you mind rebasing over main and moving this entry up to the `unreleaed` section (it's currently in the section of the last release).

pitag-ha reviewed (commented):
Perfect, thanks! Let's move the changelog entry to the `unreleased` section and merge this.

pitag-ha reviewed (approved):
Perfect, thank you!
